lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3875,25 +3904,36 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   /// The parent VJP emitter.
   VJPEmitter &vjpEmitter;
 
-  /// Dominance info for the original function.
-  DominanceInfo *domInfo;
+  /// Post-order info for the original function.
+  PostOrderFunctionInfo *postOrderInfo = nullptr;
 
-  /// Post-dominance info for the original function.
-  PostDominanceInfo *postDomInfo;
-
-  /// Mapping from original values to their corresponding adjoint values.
-  DenseMap<SILValue, AdjointValue> valueMap;
+  /// Mapping from original basic blocks to a mapping from original values to
+  /// their corresponding adjoint values.
+  DenseMap<SILBasicBlock *, DenseMap<SILValue, AdjointValue>> valueMap;
```

Comment:
Unless you need to index into this map using a BB and iterate over the nested map's keys and values, I'd suggest using a tuple (`std::pair<SILBasicBlock *, SILValue>`) as the key.

Plus, how is the "original basic block" and "original value" related? Is the value expected to be defined in the BB?

Reply:
Using `std::pair<SILBasicBlock *, SILValue>` as the key seems fine to me.

> Plus, how is the "original basic block" and "original value" related? Is the value expected to be defined in the BB?

The original value is expected to be dominated (not necessarily used) in the original basic block. For example, there may exist a key `(bb3, %0)` even though `%0` is defined in `bb0`.

Reply:
Conclusion: `DenseMap<std::pair<SILBasicBlock *, SILValue>, AdjointValue>` is the new key type. This is more efficient than `DenseMap<SILBasicBlock *, DenseMap<SILValue, AdjointValue>>`.